<a href="https://colab.research.google.com/github/LailaBulh/Procesamiento_Lenguaje_Natural/blob/main/Representacion_textos_263166.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Nombre:** Laila Montserrat Bulhosen Ramos **Matricula:** 263166

**Docente:** Dr. Rogelio Florencia Juarez

**Materia:** Procesamiento de Lenguaje Natural

**Objetivo:** Ejercicios de representación de texto

**Link Github:** [Representacion de texto](https://github.com/LailaBulh/Procesamiento_Lenguaje_Natural/blob/main/Representacion_textos_263166.ipynb)

**Fecha:** 6 marzo 2026

In [1]:
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd

In [2]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

In [3]:

!pip install unidecode

In [4]:
from unidecode import unidecode

In [5]:
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [6]:
corpus_es = [
    "Me encantó la película, los actores fueron increíbles",
    "No me gustó la película, el guion fue muy malo",
    "La actuación fue excelente, pero la historia era predecible",
    "Película aburrida, me dormí a la mitad",
    "¡Maravillosa! Recomiendo esta película a todos mis amigos",
    "El guion estaba mal escrito, pero la actuación salvó la película",
    "Demasiado larga y lenta, no la volvería a ver",
    "Me gustó mucho, los efectos especiales fueron impresionantes",
    "No es mala, pero esperaba algo más emocionante",
    "Una obra maestra del cine español, increíble dirección",
    "El final fue confuso, no entendí nada",
    "Excelente cinematografía, pero el guion flojo",
    "Demasiado predecible, ya sabía lo que iba a pasar",
    "Me reí mucho, muy divertida y entretenida",
    "La música fue espectacular, pero los actores no convencieron",
    "Película mediocre, no aporta nada nuevo",
    "Excelente historia, emocionante hasta el final",
    "No me gustó la ambientación, parecía de bajo presupuesto",
    "Muy buena dirección y fotografía, la recomiendo",
    "El guion tenía agujeros, pero los efectos visuales impresionan",
    "Aburrida, diálogo poco natural",
    "Gran actuación de los protagonistas, realmente me impactó",
    "El ritmo es lento, pero la historia es interesante",
    "No la recomiendo, desperdicié mi tiempo",
    "Un clásico moderno, me encantó cada escena",
    "Las escenas de acción fueron espectaculares",
    "El humor es pobre y los personajes poco creíbles",
    "Me emocioné, la historia me llegó al corazón",
    "La trama es confusa y difícil de seguir",
    "Película fantástica, muy bien lograda"
]


### **Ejercicio 1**

Aplicar CountVectorizer y TfidfVectorizer a corpus_es sin procesar el texto.

**CountVectorizer**

In [7]:
### Convertir a minusculas y quitar acentos
vectorizer1 = CountVectorizer(lowercase = False, strip_accents = None)

X = vectorizer1.fit_transform(corpus_es)

In [8]:
print(X)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 220 stored elements and shape (30, 136)>
  Coords	Values
  (0, 8)	1
  (0, 49)	1
  (0, 83)	1
  (0, 112)	1
  (0, 90)	1
  (0, 17)	1
  (0, 70)	1
  (0, 81)	1
  (1, 83)	1
  (1, 112)	1
  (1, 10)	1
  (1, 95)	1
  (1, 72)	1
  (1, 46)	1
  (1, 71)	1
  (1, 69)	1
  (1, 102)	1
  (1, 94)	1
  (2, 83)	1
  (2, 69)	1
  (2, 5)	1
  (2, 18)	1
  (2, 64)	1
  (2, 113)	1
  (2, 74)	1
  :	:
  (26, 116)	1
  (26, 75)	1
  (26, 115)	1
  (26, 114)	1
  (26, 36)	1
  (27, 8)	1
  (27, 83)	1
  (27, 95)	1
  (27, 74)	1
  (27, 48)	1
  (27, 87)	1
  (27, 20)	1
  (27, 35)	1
  (28, 5)	1
  (28, 53)	1
  (28, 37)	1
  (28, 131)	1
  (28, 32)	1
  (28, 40)	1
  (28, 127)	1
  (29, 102)	1
  (29, 11)	1
  (29, 65)	1
  (29, 26)	1
  (29, 89)	1


In [9]:
### Convertir arreglo a dataframe
bow_df = pd.DataFrame(X.toarray(), columns = vectorizer1.get_feature_names_out())

bow_df_matriz = bow_df.to_numpy()

display(bow_df)

,Aburrida,Demasiado,El,Excelente,Gran,La,Las,Maravillosa,Me,Muy,...,salvó,seguir,tenía,tiempo,todos,trama,ver,visuales,volvería,ya
0,0,0,0,0,0,0,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,1,0,0,...,0,0,0,0,1,0,0,0,0,0
5,0,0,1,0,0,0,0,0,0,0,...,1,0,0,0,0,0,0,0,0,0
6,0,1,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,1,0,1,0
7,0,0,0,0,0,0,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0
8,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
9,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [10]:
print(f'Dimensiones de la matriz de CountVectorizer: {bow_df.shape}')

Dimensiones de la matriz de CountVectorizer: (30, 136)


In [11]:
### Primeros 10 tokens y sus índices
all_tokens = vectorizer1.get_feature_names_out()

#print(all_tokens)

for i, t in enumerate(all_tokens):
  if i >= 10:
    break
  print(f'Token {i}: {t}')


Token 0: Aburrida
Token 1: Demasiado
Token 2: El
Token 3: Excelente
Token 4: Gran
Token 5: La
Token 6: Las
Token 7: Maravillosa
Token 8: Me
Token 9: Muy


**TfidVectorizer**

In [12]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [13]:
### Implementar TF-IDF
vectorizer2 = TfidfVectorizer(lowercase=False, strip_accents=None)

X2 = vectorizer2.fit_transform(corpus_es)

In [14]:
### Convertir a dataframe
tfid_df = pd.DataFrame(X2.toarray(), columns = vectorizer2.get_feature_names_out())

tfid_df_matriz = tfid_df.to_numpy()

display(tfid_df)

,Aburrida,Demasiado,El,Excelente,Gran,La,Las,Maravillosa,Me,Muy,...,salvó,seguir,tenía,tiempo,todos,trama,ver,visuales,volvería,ya
0,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.334091,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
1,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
2,0.000000,0.000000,0.000000,0.000000,0.000000,0.334551,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
3,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
4,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.390134,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.390134,0.000000,0.000000,0.000000,0.000000,0.000000
5,0.000000,0.000000,0.253148,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.358404,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
6,0.000000,0.375551,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.421205,0.000000,0.421205,0.000000
7,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.310067,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
8,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
9,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000


In [15]:
print(f'Dimensiones de la matriz de TfidfVectorizer: {tfid_df.shape}')

Dimensiones de la matriz de TfidfVectorizer: (30, 136)


In [16]:
### Primeros 10 tokens y sus índices
all_tokens2 = vectorizer2.get_feature_names_out()

for i, t in enumerate(all_tokens2):
  if i >= 10:
    break
  print(f'Token {i}: {t}')

Token 0: Aburrida
Token 1: Demasiado
Token 2: El
Token 3: Excelente
Token 4: Gran
Token 5: La
Token 6: Las
Token 7: Maravillosa
Token 8: Me
Token 9: Muy


In [29]:
index = 1

for doc in bow_df_matriz:
  print(f'row {index}:',doc)
  index += 1


row 1: [0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0
 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
row 2: [0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0
 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
row 3: [0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
row 4: [0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0

In [30]:
### Matriz TF IDF
index = 1

for doc in tfid_df_matriz:
  print(f'row {index}:',doc)
  index += 1


row 1: [0.         0.         0.51596817 0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.51596817 0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.57869188 0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.3642381  0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.        ]
row 2: [0.         0.         0.         0.         0.        

Con la matriz obtenida en CountVectorizer podemos observar la recurrencia de las palabras en cada oración mientras la matriz TF IDF asigna una ponderación a las palabras dependiendo la frecuencia de las palabras, las palabras más repetidas TF IDF las califica más bajas a diferencia de las palabras menos frecuentes, lo que difiere del método CountVectorizer.

Ya que ambos metodos usan los mismos datos la dimensión de las matrices es la misma.

### **Ejercicio 2**

Prepocesar el texto en corpus_es eliminando acentos, stopswords y convirtiendo a minusculas para aplicar CountVectorizer y TfidfVectorizer

In [19]:
### Función para convertir minúsculas, quitar stopwords y quitar acentos

def remove_stopwords(text):
  text = text.lower()
  stopwords_es = stopwords.words('spanish')
  tokens = word_tokenize(text, language = 'spanish')
  clean_text = [word for word in tokens if word not in stopwords_es and word.isalpha()]
  clean_text_all = ' '.join(clean_text)
  return unidecode(clean_text_all)


In [20]:
### Convertir a minusculas y quitar acentos
vectorizer_min = CountVectorizer(lowercase = True, preprocessor = remove_stopwords)

X_min = vectorizer_min.fit_transform(corpus_es)

In [21]:
print(X_min)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 123 stored elements and shape (30, 94)>
  Coords	Values
  (0, 30)	1
  (0, 76)	1
  (0, 2)	1
  (0, 56)	1
  (1, 76)	1
  (1, 48)	1
  (1, 47)	1
  (1, 66)	1
  (2, 3)	1
  (2, 41)	1
  (2, 49)	1
  (2, 79)	1
  (3, 76)	1
  (3, 0)	1
  (3, 26)	1
  (3, 68)	1
  (4, 76)	1
  (4, 83)	1
  (4, 6)	1
  (5, 76)	1
  (5, 47)	1
  (5, 3)	1
  (5, 64)	1
  (5, 35)	1
  (5, 87)	1
  :	:
  (23, 89)	1
  (24, 30)	1
  (24, 14)	1
  (24, 69)	1
  (24, 11)	1
  (24, 33)	1
  (25, 34)	1
  (25, 1)	1
  (25, 39)	1
  (26, 50)	1
  (26, 78)	1
  (26, 77)	1
  (26, 19)	1
  (27, 49)	1
  (27, 29)	1
  (27, 61)	1
  (27, 18)	1
  (28, 90)	1
  (28, 15)	1
  (28, 23)	1
  (28, 88)	1
  (29, 76)	1
  (29, 42)	1
  (29, 9)	1
  (29, 62)	1


In [22]:
### Convertir a dataframe
corpus_df = pd.DataFrame(X_min.toarray(), columns = vectorizer_min.get_feature_names_out())

display(corpus_df)

,aburrida,accion,actores,actuacion,agujeros,ambientacion,amigos,aporta,bajo,bien,...,rei,ritmo,sabia,salvo,seguir,tiempo,trama,ver,visuales,volveria
0,0,0,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,1,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5,0,0,0,1,0,0,0,0,0,0,...,0,0,0,1,0,0,0,0,0,0
6,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,1,0,1
7,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
8,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
9,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [23]:
print(f'Dimensiones de la matriz de CountVectorizer: {corpus_df.shape}')

Dimensiones de la matriz de CountVectorizer: (30, 94)


In [24]:
all_tokens3 = vectorizer_min.get_feature_names_out()

for i, t in enumerate(all_tokens3):
  if i >= 10:
    break
  print(f'Token {i}: {t}')

Token 0: aburrida
Token 1: accion
Token 2: actores
Token 3: actuacion
Token 4: agujeros
Token 5: ambientacion
Token 6: amigos
Token 7: aporta
Token 8: bajo
Token 9: bien


**TfidVectorizer**

In [25]:
### Implementar TF-IDF
vectorizer_tfidf = TfidfVectorizer(lowercase = True, strip_accents = 'unicode',preprocessor = remove_stopwords)

X2_tfidf = vectorizer_tfidf.fit_transform(corpus_es)

In [26]:
### Convertir a dataframe
tfid_df2 = pd.DataFrame(X2_tfidf.toarray(), columns = vectorizer_tfidf.get_feature_names_out())

tfid_df_matriz = tfid_df2.to_numpy()

display(tfid_df2)

,aburrida,accion,actores,actuacion,agujeros,ambientacion,amigos,aporta,bajo,bien,...,rei,ritmo,sabia,salvo,seguir,tiempo,trama,ver,visuales,volveria
0,0.000000,0.00000,0.515968,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.000000
1,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.000000
2,0.000000,0.00000,0.000000,0.496498,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.000000
3,0.499118,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.000000
4,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.696748,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.000000
5,0.000000,0.00000,0.000000,0.378626,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.464738,0.0,0.000000,0.0,0.000000,0.000000,0.000000
6,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.456675,0.000000,0.456675
7,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.000000
8,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.000000
9,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.000000


In [27]:
print(f'Dimensiones de la matriz de TfidfVectorizer: {tfid_df2.shape}')

Dimensiones de la matriz de TfidfVectorizer: (30, 94)


In [28]:
all_tokens4 = vectorizer_tfidf.get_feature_names_out()

for i, t in enumerate(all_tokens4):
  if i >= 10:
    break
  print(f'Token {i}: {t}')

Token 0: aburrida
Token 1: accion
Token 2: actores
Token 3: actuacion
Token 4: agujeros
Token 5: ambientacion
Token 6: amigos
Token 7: aporta
Token 8: bajo
Token 9: bien


Antes de aplicar CountVectorizer y TfidfVectorizer se utilizó una función para remover stopwords, convertir el texto a minúsculas y quitar acentos, después se aplicaron las funciones mencionadas anteriormente, lo que redujo la dimension de las matrices a 94 renglones vs los 136 que se obtuvieron en el ejercicio 1 sin procesar el texto.